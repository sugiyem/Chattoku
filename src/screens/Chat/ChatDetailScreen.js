import React, { useEffect, useState } from "react";
import { Alert, LogBox } from "react-native";
import { ChatContainer } from "../../styles/ChatStyles";
import { firebase } from "../../services/Firebase/Config";
import { chatType } from "../../constants/Chat";
import { fetchPrivateChatMessages } from "../../services/Chat/FetchChatMessages";
import {
  isBlockedByCurrentUser,
  isCurrentUserBlocked
} from "../../services/Friend/HandleBlockedUser";
import FetchUserInfo from "../../services/Profile/FetchUserInfo";
import ChatSections from "../../components/Chat/ChatSections";
import ChatHeader from "../../components/Chat/ChatHeader";
import { useIsFocused } from "@react-navigation/native";

const initialState = {
  username: "",
  img: ""
};

// Ignore warnings from Animated (Because of Gifted Chat)
LogBox.ignoreLogs(["Animated"]);

const ChatDetailScreen = ({ navigation, route }) => {
  const userID = firebase.auth().currentUser.uid;
  const recipientData = route.params.userData;
  const recipientID = recipientData.id;

  const [userInfo, setUserInfo] = useState(initialState);
  const [messages, setMessages] = useState([]);
  const [isGetBlocked, setIsGetBlocked] = useState(false);
  const [isBlocking, setIsBlocking] = useState(false);
  //otherUserID is used to optimize firebase use
  const [otherUserID, setOtherUserID] = useState("");
  const isFocused = useIsFocused();

  useEffect(() => {
    return FetchUserInfo({
      onSuccesfulFetch: setUserInfo,
      onFailure: (error) => Alert.alert(error.message)
    });
  }, []);

  useEffect(() => {
    if (!isFocused) return;
    if (otherUserID === recipientID) return;

    console.log("triggered");
    setOtherUserID(recipientID);

    return fetchPrivateChatMessages({
      recipientID: recipientID,
      onSuccesfulFetch: setMessages,
      onFailure: (error) => Alert.alert(error.message)
    });
  }, [isFocused]);

  useEffect(() => {
    return isBlockedByCurrentUser(
      recipientID,
      () => setIsBlocking(true),
      () => setIsBlocking(false)
    );
  }, []);

  useEffect(() => {
    return isCurrentUserBlocked(
      recipientID,
      () => setIsGetBlocked(true),
      () => setIsGetBlocked(false)
    );
  }, []);

  return (
    <ChatContainer>
      <ChatHeader
        type={chatType.PRIVATE_CHAT}
        item={recipientData}
        navigation={navigation}
      />

      <ChatSections
        type={chatType.PRIVATE_CHAT}
        userData={{ ...userInfo, id: userID }}
        receiverID={recipientID}
        messages={messages}
        updateMessages={setMessages}
        isBlocking={isBlocking}
        isGetBlocked={isGetBlocked}
      />
    </ChatContainer>
  );
};

export default ChatDetailScreen;
