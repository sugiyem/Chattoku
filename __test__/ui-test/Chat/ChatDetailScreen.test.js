import React from "react";
import renderer from "react-test-renderer";
import ChatDetailScreen from "../../../src/screens/Chat/ChatDetailScreen";

jest.mock("../../../src/services/Friend/HandleBlockedUser");
jest.mock("../../../src/services/Friend/FetchFriendStatus");
jest.mock("../../../src/services/Friend/FriendshipStatus");
jest.mock("../../../src/services/Friend/HandleGroup");
jest.mock("../../../src/services/Chat/FetchChatMessages");
jest.mock("../../../src/services/Profile/FetchUserInfo");
jest.mock("@react-navigation/native", () => ({
  useIsFocused: jest.fn()
}));

const mockNavigation = {
  goBack: jest.fn(),
  navigate: jest.fn(),
  push: jest.fn(),
  replace: jest.fn()
};

const mockRoute = {
  params: {
    userData: {
      id: "yem456",
      username: "Yemima",
      bio: "CS newbie",
      img: "yemima-img"
    }
  }
};

describe("Test Private Chat Page UI", () => {
  it("Renders correctly", () => {
    const tree = renderer
      .create(
        <ChatDetailScreen navigation={mockNavigation} route={mockRoute} />
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });
});
