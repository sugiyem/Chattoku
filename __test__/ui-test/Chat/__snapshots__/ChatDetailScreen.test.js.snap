// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Private Chat Page UI Renders correctly 1`] = `
<ViewManagerAdapter_ExpoLinearGradient
  proxiedProperties={
    Object {
      "colors": Array [
        4287090411,
        4283256319,
      ],
      "endPoint": Array [
        1,
        1,
      ],
      "locations": undefined,
      "startPoint": Array [
        0,
        0,
      ],
    }
  }
  style={
    Object {
      "flex": 1,
      "padding": 5,
    }
  }
>
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
          "flexBasis": 0,
          "flexGrow": 1,
          "flexShrink": 1,
          "justifyContent": "center",
          "width": 750,
        },
      ]
    }
  >
    <Text
      collapsable={false}
      style={
        Object {
          "color": "darkblue",
          "fontSize": 24,
          "fontWeight": "300",
          "opacity": 1,
        }
      }
    >
      Loading
    </Text>
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "flexDirection": "row",
            "justifyContent": "center",
          },
        ]
      }
    >
      <View
        collapsable={false}
        style={
          Object {
            "backgroundColor": "darkblue",
            "borderRadius": 100,
            "height": 25,
            "margin": 30,
            "padding": 5,
            "transform": Array [
              Object {
                "scaleX": 1,
              },
              Object {
                "scaleY": 1,
              },
            ],
            "width": 25,
          }
        }
      />
      <View
        collapsable={false}
        style={
          Object {
            "backgroundColor": "darkblue",
            "borderRadius": 100,
            "height": 25,
            "margin": 30,
            "padding": 5,
            "transform": Array [
              Object {
                "scaleX": 1,
              },
              Object {
                "scaleY": 1,
              },
            ],
            "width": 25,
          }
        }
      />
      <View
        collapsable={false}
        style={
          Object {
            "backgroundColor": "darkblue",
            "borderRadius": 100,
            "height": 25,
            "margin": 30,
            "padding": 5,
            "transform": Array [
              Object {
                "scaleX": 1,
              },
              Object {
                "scaleY": 1,
              },
            ],
            "width": 25,
          }
        }
      />
    </View>
  </View>
</ViewManagerAdapter_ExpoLinearGradient>
`;
